crates/generator/src/backend_manage.rs

```diff
@@ -17,72 +17,107 @@ pub struct Backend {
+        }
+        // inherit backends
+        let mut backends = self
+            .backend_switches
+            .last()
+            .map(|(_height, backends)| backends)
+            .cloned()
+            .unwrap_or_default();
+
+        // register backends
+        for config in config.backends {
+            let BackendConfig {
+                validator_path,
+                generator_path,
+                validator_script_type_hash,
+                backend_type,
+            } = config;
+            let validator = fs::read(&validator_path)
+                .with_context(|| {
+                    format!("load validator from {}", validator_path.to_string_lossy())
+                })?
+                .into();
+            let generator = fs::read(&generator_path)
+                .with_context(|| {
+                    format!("load generator from {}", generator_path.to_string_lossy())
+                })?
+                .into();
+            let validator_script_type_hash = {
+                let hash: [u8; 32] = validator_script_type_hash.into();
+                hash.into()
+            };
+            let backend = Backend {
+                validator,
+                generator,
+                validator_script_type_hash,
+                backend_type,
+            };
+            #[cfg(has_asm)]
+            if compile {
+                self.compile_backend(&backend);
+            }
+            backends.insert(backend.validator_script_type_hash, backend);
+        }
+
+        self.backend_switches.push((config.switch_height, backends));
         Ok(())
     }
 
-    pub fn register_backend(&mut self, backend: Backend) {
-        #[cfg(has_asm)]
-        {
-            self.aot_codes.0.insert(
-                backend.validator_script_type_hash,
-                self.aot_compile(&backend.generator, 0)
-                    .expect("Ahead-of-time compile"),
-            );
-            self.aot_codes.1.insert(
-                backend.validator_script_type_hash,
-                self.aot_compile(&backend.generator, 1)
-                    .expect("Ahead-of-time compile"),
-            );
-        }
+    #[cfg(has_asm)]
+    fn compile_backend(&mut self, backend: &Backend) {
+        self.aot_codes.0.insert(
+            backend.validator_script_type_hash,
+            self.aot_compile(&backend.generator, 0)
+                .expect("Ahead-of-time compile"),
+        );
+        self.aot_codes.1.insert(
+            backend.validator_script_type_hash,
+            self.aot_compile(&backend.generator, 1)
+                .expect("Ahead-of-time compile"),
+        );
+    }
 
-        self.backends
-            .insert(backend.validator_script_type_hash, backend);
+    pub fn get_backends_at_height(
+        &self,
+        block_number: u64,
+    ) -> Option<&(u64, HashMap<H256, Backend>)> {
+        self.backend_switches
+            .iter()
+            .rev()
+            .find(|(height, _)| block_number >= *height)
```

Comment:
[Binary search](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.binary_search_by_key) could be used to improve performance if the number of switches is high.

Reply:
Right, but since the chain is keep growing, the later version of backend has more possibility to be selected.